core/src/main/java/org/bitcoinj/wallet/AuthenticationKeyChain.java

```diff
@@ -42,6 +46,121 @@
     int currentIndex;
     int issuedKeys;
 
+    public static class Builder<T extends Builder<T>> {
+        protected SecureRandom random;
+        protected int bits = DeterministicSeed.DEFAULT_SEED_ENTROPY_BITS;
+        protected String passphrase;
+        protected long creationTimeSecs = 0;
+        protected byte[] entropy;
+        protected DeterministicSeed seed;
+        protected boolean isFollowing = false;
+        protected DeterministicKey spendingKey = null;
+        protected ImmutableList<ChildNumber> accountPath = null;
+        protected KeyChainType type = null;
+
+        protected Builder() {
+        }
+
+        @SuppressWarnings("unchecked")
+        protected T self() {
+            return (T)this;
+        }
+
+        /**
+         * Creates a deterministic key chain starting from the given seed. All keys yielded by this chain will be the same
+         * if the starting seed is the same.
+         */
+        public T seed(DeterministicSeed seed) {
+            this.seed = seed;
+            return self();
+        }
+
+        /**
+         * Generates a new key chain with entropy selected randomly from the given {@link SecureRandom}
+         * object and of the requested size in bits.  The derived seed is further protected with a user selected passphrase
+         * (see BIP 39).
+         * @param random the random number generator - use new SecureRandom().
+         * @param bits The number of bits of entropy to use when generating entropy.  Either 128 (default), 192 or 256.
+         */
+        public T random(SecureRandom random, int bits) {
+            this.random = random;
+            this.bits = bits;
+            return self();
+        }
+
+        /**
+         * Generates a new key chain with 128 bits of entropy selected randomly from the given {@link SecureRandom}
+         * object.  The derived seed is further protected with a user selected passphrase
+         * (see BIP 39).
+         * @param random the random number generator - use new SecureRandom().
+         */
+        public T random(SecureRandom random) {
+            this.random = random;
+            return self();
+        }
+
+        /**
+         * Creates a key chain that can spend from the given account key.
+         */
+        public T spend(DeterministicKey accountKey) {
+            checkState(accountPath == null, "either spend or accountPath");
+            this.spendingKey = accountKey;
+            this.isFollowing = false;
+            return self();
+        }
+
+        /** The passphrase to use with the generated mnemonic, or null if you would like to use the default empty string. Currently must be the empty string. */
+        public T passphrase(String passphrase) {
+            // FIXME support non-empty passphrase
```

Comment:
Missed FIXME?

Reply:
This is a copy/paste from DeterministicKeyChain.  We will need to leave it until upstream (bitcoinj fixes this).  Unfortunately, since I copy pasted this code, now there are three copies of it in our dashj library:  DeterministicKeyChain, FriendKeyChain and AuthenticationKeyChain (here).